bob-apps/bin/config_cluster_generator/pattern.rs

```diff
@@ -0,0 +1,250 @@
+        Err(anyhow!("Failed to parse the sample {}", sample))
+    }
+}
+
+fn substitute_node_pattern(node_pattern: &str, ip: &str, port: u16, id: usize) -> String {
+    let re = Regex::new(r"\{(\w+)}").unwrap();
+    re.replace_all(node_pattern, |caps: &Captures| match &caps[1] {
+        "ip" => ip.to_string(),
+        "port" => port.to_string(),
+        "id" => id.to_string(),
+        _ => caps[0].to_string(),
+    })
+    .to_string()
+}
+
+fn generate_range_samples(pattern: &str) -> impl Iterator<Item = String> {
+    let re = Regex::new(r"\[(\d+)-(\d+)]").unwrap();
+
+    let ranges = re.captures_iter(pattern).map(|captures| {
+        let start: usize = captures[1].parse().unwrap();
+        let end: usize = captures[2].parse().unwrap();
+        start..=end
+    });
+
+    re.split(pattern)
+        .zip_longest(ranges)
+        .map(|x| {
+            if let itertools::EitherOrBoth::Both(part, range) = x {
+                range.map(|i| part.to_string() + &i.to_string()).collect()
+            } else {
+                vec![x.left().expect("is some because split > range").to_string()]
+            }
+        })
+        .multi_cartesian_product()
+        .map(|x| x.concat())
+        .into_iter()
+}
+
+fn pattern_extend_disks(node: &mut Node, disk_paths: impl Iterator<Item = String>) {
+    let old_disks = node.disks();
+    let new_disks: Vec<DiskPath> = disk_paths
+        .filter(|disk_path| !old_disks.iter().any(|d| d.path() == *disk_path))
+        .enumerate()
+        .map(|(idx, disk_path)| {
+            DiskPath::new(
+                DiskName::new(&format!("disk{}", idx + old_disks.len() + 1)),
+                disk_path.as_str(),
+            )
+        })
+        .collect();
+    node.disks_extend(new_disks);
+}
+
+pub fn pattern_extend_nodes(
+    mut old_nodes: Vec<Node>,
+    pattern: String,
+    node_pattern: String,
+) -> AnyResult<Vec<Node>> {
+    let parsed_samples = generate_range_samples(&pattern)
+        .map(|key| parse_address_sample(&key))
+        .collect::<AnyResult<Vec<_>>>()?;
+    parsed_samples
+        .iter()
+        .group_by(|(ip_port, _)| ip_port)
+        .into_iter()
+        .for_each(|(ip_port, addresses)| {
+            let address = format!("{}:{}", ip_port.0, ip_port.1);
+            let mut p_node = old_nodes.iter_mut().find(|node| node.address() == address);
+            if p_node.is_none() {
+                old_nodes.push(Node::new(
+                    substitute_node_pattern(
+                        &node_pattern,
+                        &ip_port.0,
+                        ip_port.1,
+                        old_nodes.len() + 1,
+                    ),
+                    address,
+                    vec![],
+                ));
+                p_node = old_nodes.last_mut();
+            }
+            let p_node = p_node.expect("is some because it's either the pushed one or found");
+            pattern_extend_disks(p_node, addresses.map(|(_, path)| path.to_owned()));
```

Comment:
I think this cycle should be split in two: 
1. Create missing nodes
2. Extend disks for all nodes